nixos/lib/make-zfs-image.nix

```diff
@@ -0,0 +1,340 @@
+{ lib
+, pkgs
+, # The NixOS configuration to be installed onto the disk image.
+  config
+
+, # The size of the root disk, in megabytes.
+  rootSize ? 2048
+
+, # The name of the ZFS pool
+  bootPoolName ? "boot"
+
+, # size of the boot disk
+  bootSize ? 1024
+, # features on the boot pool
+  # Note: the pool has no features enabled by default,
+  # and only this list of features is enabled. This list
+  # must match GRUB's supported list.
+  # See: https://github.com/openzfs/openzfs-docs/blame/master/docs/Getting%20Started/Debian/Debian%20Stretch%20Root%20on%20ZFS.rst#L200-L216
+  bootPoolFeatures ? [
+    "async_destroy"
+    "bookmarks"
+    "embedded_data"
+    "empty_bpobj"
+    "enabled_txg"
+    "extensible_dataset"
+    "filesystem_limits"
+    "hole_birth"
+    "large_blocks"
+    "lz4_compress"
+    "spacemap_histogram"
+    "userobj_accounting"
+  ]
+
+, # zpool properties
+  bootPoolProperties ? {
+    autoexpand = "on";
+  }
+, # pool-wide filesystem properties
+  bootPoolFilesystemProperties ? {
+    acltype = "posixacl";
+    atime = "off";
+    compression = "on";
+    mountpoint = "legacy";
+    xattr = "sa";
+  }
+, # The name of the ZFS pool
+  rootPoolName ? "tank"
+
+, # zpool properties
+  rootPoolProperties ? {
+    autoexpand = "on";
+  }
+, # pool-wide filesystem properties
+  rootPoolFilesystemProperties ? {
+    acltype = "posixacl";
+    atime = "off";
+    compression = "on";
+    mountpoint = "legacy";
+    xattr = "sa";
+  }
+
+, # datasets, with per-attribute options:
+  # mount: (optional) mount point in the VM
+  # properties: (optional) ZFS properties on the dataset, like filesystemProperties
+  # Note: datasets will be created from shorter to longer names as a simple topo-sort
+  datasets ? {
+    "${rootPoolName}/system/root".mount = "/";
+    "${rootPoolName}/system/var".mount = "/var";
+    "${rootPoolName}/local/nix".mount = "/nix";
+    "${rootPoolName}/user/home".mount = "/home";
+    "${bootPoolName}/boot".mount = "/boot";
+  }
```

Comment:
These really scream "we should be evaluated through a small modules system implementation".

Overriding only one option is harder than it should be.

Reply:
We could delete all these and require the user to specify them. The only existing call site now _does_ use the module system.

Reply:
I would tend to agree here. Especially since it also better demonstrates "this is for internal use, external users have to do so much more to make it work just right at the moment".

Reply:
Good idea, thanks. Done.